Firestore/core/test/firebase/firestore/util/async_queue_test.cc

```diff
@@ -0,0 +1,246 @@
+  auto signal_finished = CreateSignal();
+
+  auto queue = Queue();
+  queue.Enqueue([&] {  // clang-format off
+    EXPECT_ANY_THROW(queue.Enqueue([&] { signal_finished(); }););
+    // clang-format on
+  });
+
+  WaitForTestToFinish(&signal_finished);
+}
+
+TEST(AsyncQueue, EnqueueAllowsEnqueuedTasksToUseEnqueueUsingSameQueue) {
+  auto signal_finished = CreateSignal();
+
+  auto queue = Queue();
+  queue.Enqueue([&] {  // clang-format off
+    queue.EnqueueAllowingSameQueue([&] { signal_finished(); });
+    // clang-format on
+  });
+
+  EXPECT_TRUE(WaitForTestToFinish(&signal_finished));
+}
+
+TEST(AsyncQueue, SameQueueIsAllowedForUnownedActions) {
+  auto signal_finished = CreateSignal();
+  auto queue = Queue();
+
+  using WrapT = std::pair<AsyncQueue*, std::packaged_task<void()>*>;
+  WrapT wrap{&queue, &signal_finished};
+  dispatch_async_f(underlying_queue, &wrap, [](void* const raw_wrap) {
+    auto unwrap = static_cast<const WrapT*>(raw_wrap);
+    unwrap->first->Enqueue([unwrap] { (*unwrap->second)(); });
+  });
+
+  EXPECT_TRUE(WaitForTestToFinish(&signal_finished));
+}
+
+TEST(AsyncQueue, RunSync) {
+  bool finished = false;
+
+  auto queue = Queue();
+  queue.RunSync([&] { finished = true; });
+
+  EXPECT_TRUE(finished);
+}
+
+TEST(AsyncQueue, RunSyncDisallowsEnqueuedTasksToUseEnqueue) {
+  auto queue = Queue();
+  queue.RunSync([&] {  // clang-format off
+    EXPECT_ANY_THROW(queue.RunSync([] {}););
+    // clang-format on
+  });
+}
+
+TEST(AsyncQueue, EnterCheckedOperationDisallowsNesting) {
+  auto queue = Queue();
+  queue.RunSync(
+      [&] { EXPECT_ANY_THROW(queue.EnterCheckedOperation([&] {});); });
+}
+
+TEST(AsyncQueue, VerifyIsCurrentQueueRequiresCurrentQueue) {
+  ASSERT_NE(underlying_queue, dispatch_get_main_queue());
+  EXPECT_ANY_THROW(Queue().VerifyIsCurrentQueue());
+}
+
+TEST(AsyncQueue, VerifyIsCurrentQueueRequiresOperationInProgress) {
+  auto queue = Queue();
+  dispatch_sync_f(underlying_queue, &queue, [](void* const raw_queue) {
+    EXPECT_ANY_THROW(
+        static_cast<AsyncQueue*>(raw_queue)->VerifyIsCurrentQueue());
+  });
+}
+
+TEST(AsyncQueue, VerifyIsCurrentQueueWorksWithOperationInProgress) {
+  auto queue = Queue();
+  queue.RunSync([&] { EXPECT_NO_THROW(queue.VerifyIsCurrentQueue()); });
+}
+
+TEST(AsyncQueue, CanScheduleOperationsInTheFuture) {
+  auto signal_finished = CreateSignal();
+  std::string steps;
+
+  auto queue = Queue();
+  queue.Enqueue([&steps] { steps += '1'; });
+  queue.EnqueueAfterDelay(AsyncQueue::Milliseconds(5), kTimerId1,
+                          [&steps, &signal_finished] {
+                            steps += '4';
+                            signal_finished();
+                          });
+  queue.EnqueueAfterDelay(AsyncQueue::Milliseconds(1), kTimerId2,
+                          [&steps] { steps += '3'; });
+  queue.Enqueue([&steps] { steps += '2'; });
+
+  EXPECT_TRUE(WaitForTestToFinish(&signal_finished));
+  EXPECT_EQ(steps, "1234");
+}
```

Comment:
I don't think that's a problem, but I may misunderstand what you mean. My thinking is:
* libdispatch guarantees the relative order of the dispatched work. That is, work scheduled with 1ms delay is guaranteed (AFAICT) to be run before the work scheduled with 5ms delay. The queue the tests are using is a serial queue, so the tasks are run one at a time.
* if libdispatch takes too long, I don't see how a fake clock can help.
* I'm also not sure how exactly libdispatch checks time (to be more precise, whether it's well-documented).

So I don't think there should be a problem - the test basically checks the correct ordering, and isn't really timing-dependent (well, it can time out, but I think it's an unavoidable problem). Please let me know if I'm missing something.

Reply:
I meant the following:

```
steps = '1';
queue.Enqueue(delay=4, fn=() { steps += '4';}
queue.Enqueue(delay=3, fn=() { steps += '3';}
queue.Enqueue(delay=2, fn=() { steps += '2';}
```

Expected output would be "1234". (Good.) But now consider:

```
steps = '1';
queue.Enqueue(delay=4, fn=() { steps += '4';}
MasiveContextSwitchCausingDelayOfAtLeast5();
queue.Enqueue(delay=3, fn=() { steps += '3';}
queue.Enqueue(delay=2, fn=() { steps += '2';}
```

In this case, could the output be "1423"?

If libdispatch doesn't actually start processing things in the queue until WaitForTestToFinish(), then no action required. But if it could start processing as soon as something is added to the queue, it seems like you could end up in trouble.


> if libdispatch takes too long, I don't see how a fake clock can help.

Not due to processing, but due to sleeps. For instance, this test will take at least 5ms to run. (Which is pretty quick. But imagine if the test is actually flakey and a temporary measure to fix that causes us to increase that to 5s.) Sitting around and waiting for those timeouts/delays to expire isn't awful for integration tests, but is unfortunate for unit tests.

```
steps = '1';
queue.Enqueue(delay=4, fn=() { steps += '4';}
queue.Enqueue(delay=3, fn=() { steps += '3';}
queue.Enqueue(delay=2, fn=() { steps += '2';}
fakeClock.Advance(5);
WaitForQueueToFinish(); // still necessary, since we need to allow the queue to notice the time and run async tasks. But should complete more-or-less immediately.
```

Of course, that all assumes you *can* inject a fake clock into libdispatch. If it has a hard dependency on the system clock then there's probably not much that can reasonable be done.

Reply:
> In this case, could the output be "1423"?

It's a serial queue. Unless I have a pretty big misunderstanding of how libdispatch works, it guarantees relative order of dispatched operations. The next dispatch call on a serial queue is only ever made when the previous one has finished. So I don't see how this could be a problem.

> For instance, this test will take at least 5ms to run.

All the actual code that uses this functionality has delay of several seconds. I chose milliseconds granularity for the sake of tests pretty much arbitrarily. It could be nanoseconds, what do you think?

> Of course, that all assumes you can inject a fake clock into libdispatch. If it has a hard dependency on the system clock then there's probably not much that can reasonable be done.

I'm pretty skeptical whether it's possible, and whether it's something worth relying upon.

Reply:
> So I don't see how this could be a problem.

Lets sync in person. (I've got a repro for my concern locally. Whether it's a real concern or not...?)

> I chose milliseconds granularity for the sake of tests pretty much arbitrarily. It could be nanoseconds, what do you think?

Oh, as long as there's no race, I don't care. 5ms won't be our biggest bottleneck. :)